### PR TITLE
Makefile and other visual changes.

### LINKED ISSUE #1510: Optimise golang version update across multiple files
/area provider/ibmcloud

When a new golang version is out, it is being updated manually in 
1. [Dockerfile](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/Dockerfile#L18)
2. [hack/ccm/Dockerfile](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/hack/ccm/Dockerfile#L21)
3. [Makefile](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/Makefile#L21)

This issue is to explore this process can be optimised or automated.

### PR DESCRIPTION
**Pull request by kishen-v**
**What this PR does / why we need it**: 
Changes in this PR:
1. `make serve` is now made available from the parent Makefile and can be executed from the repository root.
2. `README.md` changes to fix the broken link of License, and other related badges, alignment of the images, etc.
3.   Version bump of `mdbook` from v0.4.5 to v0.4.35.
4.  Add `make update-go` to update golang version across multiple files.
5.  Add `make verify-go-version` to find if there are any recent release of Go that can be updated.
```
Kishens-MacBook-Pro: cluster-api-provider-ibmcloud : mdbook-target ᐅ  make update-go VERSION=1.21.4 
-FROM --platform=${BUILDPLATFORM} golang:1.20.11 as toolchain
+FROM --platform=${BUILDPLATFORM} golang:1.21.4 as toolchain

-GO_VERSION ?= 1.20.11
+GO_VERSION ?= 1.21.4

-FROM golang:1.20.11 AS ccm-builder
+FROM golang:1.21.4 AS ccm-builder
```
6.  Update favicon image to show clusterapi logo, from the default mdbook logo.
Pasted Graphic 4

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1510

**Special notes for your reviewer**: This PR has multiple commits. Please let me know if it needs to be made into separate PRs.

/area provider/ibmcloud

**Release note**:

```release-note
make serve is now made available from the parent Makefile and can be executed from the repository root.
README.md changes to fix the broken link of License, and other related badges, alignment of the images, etc.
Version bump of mdbook from v0.4.5 to v0.4.37.
Add make update-go to update golang version across Makefiles
Add make verify-go-version to find if inconsistencies exist across files with regards to go version in Make/Submakefiles.
Update favicon image to ClusterAPI logo.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -600,3 +600,21 @@ kind-cluster: ## Create a new kind cluster designed for development with Tilt
...
+.PHONY: serve
+serve: ## Build the CAPIBM book and serve it locally to validate changes in documentation.
+	make -C docs/book/ serve
+
+## --------------------------------------
+## Update Go Version
+## --------------------------------------
+
+.PHONY: update-go
+update-go: ## Update Go version across files - Usage make update-version VERSION=X.YY.ZZ
```
Comment (mkumatag) [inline Makefile:617]: can we somehow call this in the verify flow and ensure files are always updated?

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (kishen-v) [inline Makefile:617]: Really considering `make update-version X.Y.Z` than `make update-version VERSION=X.YY.ZZ`.
Will look into it.

File: versions.mk [hunk 2]
```diff
@@ -12,4 +12,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-MDBOOK_VERSION := v0.4.5
+MDBOOK_VERSION := v0.4.35
```
Comment (Prajyot-Parab) [inline versions.mk:15]:
```suggestion
MDBOOK_VERSION := v0.4.36
```
New release is out!

File: Makefile (hunk 1 above)
Comment (Prajyot-Parab) [reply Makefile:617]: @kishen-v any update?

**Review by Prajyot-Parab — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (kishen-v) [reply Makefile:639]: Hi @mkumatag, 
I had discussed with @Prajyot-Parab regarding this and have added a new target `verify-go-version` which checks if a patch release is available for the n-1'th minor release (Just the way we follow) and chose not to block PRs when the version check fails.

I have a single commit for this action in a different branch to trigger for all PR, which confirms if we're using the latest patch release.  I shall enable this action for this repository if this seems ok.
The current behavior is that a warning is raised in the workflow regarding the version mismatch.
image

along with the following output in the shell:
```
go1.20.12 is available on go.dev, the version in Makefile is 1.20.11.
Consider bumping the version, run make update-go VERSION=go1.20.12 to update across all files.
```

The above check is a part of `verify` as well. Since we do not mark the job as failed, the findings related to version remains as a few lines of log in the prow-job.

File: Makefile [hunk 3]
```diff
@@ -25,6 +25,8 @@ IMG ?= controller:latest
 # Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
 CRD_OPTIONS ?= "crd:crdVersions=v1"
 
+SHELL=/bin/bash
```
Comment (kishen-v) [inline Makefile:28]: Observing `/bin/sh: 1: Syntax error: redirection unexpected` as here-strings are used in verify-go-version.
 Hence, setting shell to bash shell from user shell as a workaround.

File: Makefile [hunk 4]
```diff
@@ -539,6 +541,26 @@ verify-security: ## Verify code and images for vulnerabilities
...
+.PHONY: verify-go-version
+verify-go-version: ## Confirms the availability of minor/patch release.
+	available_version=$$(curl -s "https://go.dev/dl/?mode=json" | jq ".[1].version" | tr -d '"'); \
+	avl_minor_patch_version=$$(echo $$available_version | awk -F'.' '{print $$(NF-1),$$(NF)}'); \
+	IFS=' ' read -r available_minor available_patch <<< "$$avl_minor_patch_version"; \
+	current_minor_patch_version=$$(echo $(GO_VERSION)| awk -F'.' '{print $$(NF-1),$$(NF)}'); \
+	IFS=' ' read -r current_minor current_patch <<< "$$current_minor_patch_version"; \
+	if [ $$current_minor -le $$available_minor ]; then \
+		if [ $$current_patch -lt $$available_patch ] || [ $$current_minor -lt $$available_minor ]; then \
+			echo "::warning ::Update GO_VERSION to $$available_version"; \
```
Comment (kishen-v) [inline Makefile:554]: Ref: https://stackoverflow.com/a/75849954
image

File: Makefile (hunk 1 above)
Comment (mkumatag) [reply Makefile:639]: I feel allowing user to decide what version they want to use, we can update the version manually or by some automation like dependatbot will be better idea.

**Comment by Prajyot-Parab** [comment]
@kishen-v 
Is this the expected behaviour?

```
prajyotparab@Prajyots-MacBook-Pro cluster-api-provider-ibmcloud % make verify-go-version
fatal: No names found, cannot describe anything.
::warning ::Update GO_VERSION to go1.20.13
go1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run make update-go VERSION=go1.20.13 to update across all files.
```

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

@Prajyot-Parab Please take a look

**Comment by kishen-v** [comment]
@Prajyot-Parab, If it's this line `fatal: No names found, cannot describe anything.` 
I see the same error on my machine. I shall look into it.

**Comment by Prajyot-Parab** [comment]
> @Prajyot-Parab, If it's this line `fatal: No names found, cannot describe anything.` I see the same error on my machine. I shall look into it.

Yes, please check.

**Comment by kishen-v** [comment]
@Prajyot-Parab, I see it now.. git describe seems to be causing a problem. 
```
Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target: git describe
fatal: No names found, cannot describe anything.
```
added `--always`  to pick the latest commit object as fallback to supress the error.
```
Kishens-MacBook-Pro: cluster-api-provider-ibmcloud : mdbook-target :  git describe --always --abbrev=0
44b176a7ffd2a0253429e29c53da1613924bbca2
```

Doc:  https://git-scm.com/docs/git-describe#Documentation/git-describe.txt---always

With the above change in the Makefile and version.sh, the error is no longer seen. 
```
Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target:  make verify-go-version
::warning ::Update GO_VERSION to go1.20.13
go1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run make update-go VERSION=go1.20.13 to update across all files.
```

**Comment by kishen-v** [comment]
@Prajyot-Parab, ptal! Required changes have been made and pushed to branch.

File: Makefile
Comment (Prajyot-Parab) [inline Makefile:1]: Need to fix:

```
::warning ::Update GO_VERSION to go1.20.13
go1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run make update-go VERSION=go1.20.13 to update across all files.
prajyotparab@Prajyots-MacBook-Pro cluster-api-provider-ibmcloud % make update-go VERSION=go1.20.13
```

it should suggest -- > ```make update-go VERSION=1.20.13```

Comment (kishen-v) [reply Makefile:1]: Thanks for pointing out! 
The changes are made to print just the version.
```
Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target :  make verify-go-version
::warning ::Update GO_VERSION to 1.20.13
1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run "make update-go VERSION=1.20.13" to update across all files.
```

File: Makefile [hunk 5]
```diff
@@ -18,13 +18,15 @@ ROOT_DIR_RELATIVE := .
 
 include $(ROOT_DIR_RELATIVE)/common.mk
 
-GO_VERSION ?= 1.20.12
```
Comment (Prajyot-Parab) [inline Makefile:21]: @kishen-v why this change?

Comment (kishen-v) [reply Makefile]: Sorry, it was an unintended change which went along with latest commit. 
I have reverted back to the version on `main`.

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

@Karthik-K-N ptal

**Review by Karthik-K-N — APPROVED** [review]
Thanks for the PR

/lgtm

**Comment by kishen-v** [comment]
Hi @Prajyot-Parab/ @Karthik-K-N,

I have incorporated the discussed changes where:
1. Do away with go version set in the Makefile.
2. Retrieve the go-version from the Dockerfile in repository root.
3. Update go-version across all Dockerfiles in the repository.

Here are the o/ps.
```
Updating versions across all dockerfiles.

Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target :  make update-go VERSION=1.20.12
Updated go version to 1.20.12 across Dockerfile hack/ccm/Dockerfile

Verify that Go versions are same across all Dockerfiles

Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target : make verify-go-version
Versions are uniform across Dockerfiles, the go-version used is 1.20.12

And if not...

Kishens-MacBook-Pro: cluster-api-provider-ibmcloud : mdbook-target  :  make verify-go-version
Versions are different across Dockerfiles, please use 'make update-go'
```

cc: @mkumatag.

**Review by mkumatag — APPROVED** [review]
I will let @Prajyot-Parab to add lgtm

**Comment by Prajyot-Parab** [comment]
@kishen-v lets do it this way
1. Refer https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/Makefile#L23-L24 ,lets use same approach here, we will pass the image itself as build-arg to our docker commands and that way no need to fetch any version from Dockerfiles.
2. Lets have check in place to verify the version mentioned in Makefile and ccm dockerfile is same, If yes all good, If no just throw out warning saying version mismatch and suggest user to have consistency.
Also you can refer this older commit (https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/commit/3e0af2a737146aff17a2b268ecd285dd8e74c992) lets switch to this in ccm Dockerfile (this will avoid triggering update via depenbot), you can also pass that image as arg via ccm Makefile and have comparison done on it rather than Dockerfile
3. Lets not have any command in place to update the versions via make  (drop that logic)

let me know incase of any queries.

**Comment by Prajyot-Parab** [comment]
@kishen-v squash everything into 1 commit. (rest lgtm)

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm
